### DISCUSSION TITLE
CLI

### DISCUSSION THREAD

**Discussion by NWBY** [confetti-framework/confetti#105] [Ideas]
Big Laravel fan here and have just discovered Confetti and really liking it so far.

I’m wondering if there are any plans to implement a CLI tool similar to Artisan to generate new Models, Migrations, etc. 

Happy to discuss, plans and contribute to this if needed.

**Comment by Wulfheart** [comment]
I'm sure that there are plans for that.
However, I'd like to know if there is a concept of service providers in packages like laravel has. This would simplify the embedding of commands.

**Reply by reindert-vetter** [reply]
Documentation of Service Providers:
https://www.confetti-framework.com/docs/architecture-concepts/providers.html

**Comment by reindert-vetter** [comment]
I have decided to focus on a CLI tool. Maybe you can help me come up with a name. See also: https://www.reddit.com/r/laravel/comments/lhg31y/do_you_know_a_name_for_a_cli_tool_for_confetti_go/

**Reply by Wulfheart** [reply]
What about Wright? It is a synonym for artisan.

**Reply by Wulfheart** [reply]
Furthermore: Yesterday I had the idea to make it like artisan: go run cliname [command]

**Reply by reindert-vetter** [reply]
Actually, I would like to omit 'go run' 🙈

**Reply by Wulfheart** [reply]
Why? This would allow to add commands dynamically.

**Reply by marcelloh** [reply]
but you need to program those "dynamic" commands too, so there's nothing dynamic about it.

**Reply by marcelloh** [reply]
Clifetti?

**Reply by reindert-vetter** [reply]
> Why? This would allow to add commands dynamically.

But of course. See another message in this discussion.

**Comment by reindert-vetter** [comment]
Working hard to make an Artisan alternative to Confetti. A preview:
![carbon (4)](https://user-images.githubusercontent.com/16726304/107978066-79077a00-6fbc-11eb-869f-e0707d3d7169.png)

**Reply by bosix** [reply]
It would be necessary to compile `main` as the binary before using the command, right?

Is it possible to have a look at the current version of the source code? I'm quite interested in how this command is exactly integrated and how is decided whether confetti is started or command should be executed.

**Reply by reindert-vetter** [reply]
The current command will change:
- `go run main.go` or` main` will list all commands:
![image](https://user-images.githubusercontent.com/16726304/108120591-d0cbe100-70a1-11eb-8ba9-d8fc36f9ca79.png)
- `go run main.go app:serve` or `main app:serve` will start the webserver.
- `go run main.go your:first-command --dry-run` or`main your:first-command --dry-run` will be a custom command.

With `go run main.go` Go will always use your current code

**Reply by reindert-vetter** [reply]
> Is it possible to have a look at the current version of the source code? I'm quite interested in how this command is exactly integrated and how is decided whether confetti is started or command should be executed.

If I have it a little more working then I'll share an online git diff with you.

**Reply by reindert-vetter** [reply]
@bosix The first version has been made. It now only works for boolean flags. In addition, I have not yet completely gone through and polished my code.

[confetti](https://github.com/confetti-framework/confetti/compare/main...%23105-baker)
[foundation](https://github.com/confetti-framework/foundation/compare/main...%23105-baker)
[contract](https://github.com/confetti-framework/contract/compare/main...%23105-baker)

I'd love to hear if you have any feedback. I will also add you as a reviewer when I open my pull request.

**Reply by bosix** [reply]
@reindert-vetter thank you for sharing. This looks very nice.

Here are a few things I want to mention:

I think the way flags are parsed into the command struct could be done easier. You already use reflection to detect the name of the flag and flagset to handle the flags but I think if we can prevent creating []ParsedOption (GetOptions() in option.go) and some of the other boilerplating:

```golang
package main

import (
	"flag"
	"fmt"
	"reflect"
)

type MyCommand struct {
	StringFlag string `short:"f" long:"flag" description:"this is how to use the command"`
}

func main() {
	cmd := reflect.TypeOf(MyCommand{})
        // I skip iterating over all fields to make the example easiert to read
	field := cmd.Field(0)
	shortName := field.Tag.Get("short")
	description := field.Tag.Get("description")

	myCmd := MyCommand{}

	switch field.Type.Kind() {
	case reflect.String:
		flag.StringVar(&myCmd.StringFlag, shortName, "", description)
	case reflect.Bool:
		//flag.BoolVar(&myCmd.StringFlag, shortName, false, description)
		panic("proper implementation needed")
	}

	flag.Parse()

	fmt.Printf("%+v", myCmd)
}
```

Even if you don't want to parse the value directly into the command you could use struct `flag.String()` which returns a pointer to the actual parsed value which can be used afterward. 
I also like that error handling and help output are already done within the flag package and we do not need any parsing since the default pattern and parsing of golang is used (I saw that your implementation works well with `-f hello` but not with `-f=hello`)

![image](https://user-images.githubusercontent.com/26360670/108621598-d7f14700-7433-11eb-99a3-783d03b26204.png)

The default/fallback value could also be set via reflection or inside the kernel when the command is registered.

It would be a pleasure to review the code when it's done. This may take some time since I'll move to a new apartment in the next few days and I don't know how much time I'll have to review.

**Reply by reindert-vetter** [reply]
> I think if we can prevent creating []ParsedOption (GetOptions() in option.go)

Thanks for your feedback. I looked at another one and the parsing data from the struct is used in 2 places: When registering the flags and setting the values. I thought it was nice to collect the data from a struct. But I'll investigate this further tomorrow. Feel free to try the code and see if you can make it more beautiful!

> first example

Thanks for your example. In your example you have a switch case for each type. The problem with that, however, is that if someone defines a custom type in a field (for example `[] model.User`) it is not easy to modify.

> I also like that error handling and help output are already done within the flag package

You are absolutely right. Thanks for your tip. I now use that in an example Getter https://github.com/confetti-framework/confetti/pull/120/files#diff-78dcb0b45b05161a9dacddf241a5ff995ef1ad290407711d4ea77fe291d37dbaR22. The error looks really good!

> I saw that your implementation works well with -f hello but not with -f=hello

Fixed. It now works with all types offered by golang itself. In addition, I made 2 examples so that people can use them as an example for self-made flag.Getter.

> The default/fallback value could also be set via reflection or inside the kernel when the command is registered.

Indeed. I still find this difficult because I would not expect it when registering. How do you feel about defining that in the field tag? For example `Username string `flag:"username" default:"bosix"`?

> New apartment

🏠 Congratulations on your new apartment! 🎊  I hope your move is going well. Take your time!

I will add extra functionalities in the meantime.

PR:
https://github.com/confetti-framework/confetti/pull/120
https://github.com/confetti-framework/foundation/pull/13

**Reply by bosix** [reply]
Just a few thoughts without looking too deep in the new version of the source code:

> Thanks for your example. In your example you have a switch case for each type. The problem with that, however, is that if someone defines a custom type in a field (for example [] model.User) it is not easy to modify.

Do you think that it's necessary to allow custom/complex types for command params? I was just thinking about how a user struct would be passed via cli: maybe as json `go run main.go process:user "{\"name\":\"john\"}"`. In my opinion, we shouldn't allow this. We could, if necessary, use `reflect.Slice` and `reflect.Struct` together with some fancy parsing to process such params but I don't think that there is any use-case for this. Any kind of switch is needed anyway since at someplace we need to decide which target datatype the input should be converted into.

> How do you feel about defining that in the field tag?

Sounds great. We should be careful not to make too much stuff with reflections and tags since the more we do with this the more can go wrong at runtime and the better all tests must be written. Maybe we could suggest the developer the "New"-pattern of go when initializing a struct:

```golang
func NewMyCommand() MyCommand {
	return MyCommand{
		StringFlag: "test",
	}
}
```

The kernel would then call the function instead of creating an empty command struct.

I'll have a look at the new changes you made hopefully later this week and maybe I'll also have some time to add some more code examples or suggestions on how to improve your code.

**Reply by reindert-vetter** [reply]
> custom/complex types for command params - it's necessary

I had the idea to pass user ids to the command. I would often use it to turn on feature flags. But you are actually right that we should not encourage complex types. If we document this, it should be at the bottom of the page.

> default values

You're right we shouldn't use more reflections. We can also choose not to implement the default values. Although I don't want to use generics too much, but this can be a nice use case to be able to register both a struct and a new function.

Thanks for your feedback! You are really right that we should not use too much magic.

**Comment by KinyaElGrande** [comment]
Is there a plan to also refactor the CLI log once one fires up the application, for example, to be displaying the link also to the application [localhost:80](http://localhost:80) and also the log to be containing all requests made to your application?

**Reply by bosix** [reply]
@KinyaElGrande that's a good idea. Currently, it's not possible to see the requests as long as you do not have any kind of reverse proxy. You would, however, use a middleware to get log messages. Here a small example:

```golang
// app/http/middleware/request_log.go
type RequestLog struct{}

func (_ RequestLog) Handle(request inter.Request, next inter.Next) inter.Response {
	resp := next(request)

	msg := fmt.Sprintf("%s on %s - status %d", request.Method(), request.Path(), resp.GetStatus())

	request.App().Log("stdout").Info(msg)

	return resp
}

// app/http/middleware/web.go
var Web = []inter.HttpMiddleware{
	middleware.DefaultResponseOutcome{Outcome: outcome.Html},
	RequestLog{},
}
```

**Reply by reindert-vetter** [reply]
Indeed I think it is better to implement yourself.

Or do you mean more that you want some sort of debugging tool? I also thought of returning a 'report id' or 'request id' with the response. So that you can easily find errors from the logs.

**Reply by bosix** [reply]
I think a request-id should always be added to any log message. That way you can find log messages which belong to the same request very easily. 

Laravel does not have such a feature by default so we have implemented it by ourselves since it's quite impossible to gain any information from log messages if every request creates multiple messages and many users are using the application at the same time. Maybe that's a good idea for a new feature for confetti.

**Reply by reindert-vetter** [reply]
Yes, we also had to build it in ourselves. I think it's a good idea to implement this in Confetti.

**Reply by KinyaElGrande** [reply]
Hehe i love the idea : ).

**Reply by reindert-vetter** [reply]
See https://github.com/confetti-framework/confetti/discussions/121

**Comment by reindert-vetter** [comment]
Today I've been working on the usage message of a command. 

Before (default Go): 
![image](https://user-images.githubusercontent.com/16726304/108918212-4d763680-7631-11eb-8755-892b20dc0ae0.png)
After:
![image](https://user-images.githubusercontent.com/16726304/108924525-1442c400-763b-11eb-83bd-d3bf2022e516.png)

Do you think we should stick to the standards? https://clig.dev/#help

**Reply by KinyaElGrande** [reply]
Yes i think we should stick to the Standards.

**Reply by bosix** [reply]
I'll have to say that I like both versions. The first one because it's the default of go and the second one because it also shows the required data types (int, string, etc.).

I think the decision depends a bit on the effort needed to implement the second version properly.

**Comment by reindert-vetter** [comment]
@bosix @KinyaElGrande @Wulfheart @NWBY 
I have to do some extra testing and I have to check the documentation, but everything should work. Feel free to take a look:
https://www.confetti-framework.com/docs/digging-deeper/commands

https://github.com/confetti-framework/confetti/releases/tag/v0.20.0

**Reply by bosix** [reply]
@reindert-vetter thank you for sharing

I'm a bit confused because any go run ends up in a panic:
![image](https://user-images.githubusercontent.com/26360670/109545812-9d3d7d80-7ac9-11eb-8b9b-c039b682329f.png)

I've downloaded the release you mentioned and ran `go get -v` and `go run main.go`. Do you have any idea where my mistake is?

**Reply by reindert-vetter** [reply]
@bosix You are absolutely right. Bug has been found and fixed.
require `github.com/confetti-framework/foundation v0.5.5` to use the fix.


https://github.com/confetti-framework/confetti/releases/tag/v0.20.1

**Reply by bosix** [reply]
Nice! It's working now.

There is one thing we should change: we should provide the help output which is printed when no command is given when the user types in an unknown command as well.
![image](https://user-images.githubusercontent.com/26360670/109700019-8f532f80-7b91-11eb-8db7-91ebbece0bc1.png)

**Reply by reindert-vetter** [reply]
I've been thinking about that. Maybe we should only do that if you have less than 10 commands. At work, we already have more than 100 commands. The best thing would be if we could guess which command it should be. There's a library for that, but I can't find it anymore. What are your thoughts on this?

<img src="https://user-images.githubusercontent.com/16726304/109712861-d4329280-7ba0-11eb-8ec3-b75948a3a9e4.png" width="400px"> <img src="https://user-images.githubusercontent.com/16726304/109713043-08a64e80-7ba1-11eb-807c-4d4fb9270842.png" width="350px">

**Reply by bosix** [reply]
You are absolutely right. Artisan prints all commands if you run `php artisan` and that's not very useful since even vanilla laravel has many commands to show up.

I know that the cli of laravel is very powerful but even if such a library does not exist for go there could be simple guessing been implemented with some usage of `strings.Contains()`. Confetti could iterate through the signature of all registered commands and check if parts of the command the user typed in are somewhere found in another command. Not the best prediction but a beginning. At work, we often type in the beginning of the command to get all sub-commands (in your example I would type `php artisan db` to get all db commands)

**Reply by reindert-vetter** [reply]
Yes, that is the best solution for the time being, I think. I haven't implemented it in the first version now, but I will definitely implement this later.

**Reply by bosix** [reply]
What do you think about this?

![image](https://user-images.githubusercontent.com/26360670/110038253-c7987080-7d3f-11eb-822b-5876a80184a4.png)

Edit: here are the necessary changes (untestet) https://github.com/confetti-framework/foundation/compare/main...bosix:main

**Reply by reindert-vetter** [reply]
Looking good! That does not look that difficult indeed. If you make a few tests, you can create a pull request. Otherwise, I will implement this next week.

**Reply by bosix** [reply]
I can also provide some tests for the code and I'm already on it. My only problem is that I've no idea how to test the confirmation. I tried to write to `kernel.Writer` but this does not work.

**Reply by reindert-vetter** [reply]
I will make situations with a Stdin more accessible for testing. Maybe I'll get it done tonight, otherwise tomorrow.

**Reply by reindert-vetter** [reply]
I was able to fix it in github.com/confetti-framework/foundation v0.5.9. From test/console/output_test.go:124 you will see a number of tests. You can use those tests for inspiration. Don't forget to add a newline when specifying the writer.

You can define the writer using NewCliByReadersAndWriter(). Bit of a long name for a constructor, but I also thought it is a bit overkill to use [callbacks options](https://gist.github.com/travisjeffery/8265ca411735f638db80e2e34bdbd3ae) for this.

Of course, feedback is welcome!

**Reply by bosix** [reply]
Well, the name is a bit clunky but the function just works like a charm. I've finished the tests and I've seen that you are already on the pr :)

**Comment by reindert-vetter** [comment]
@bosix @KinyaElGrande @Wulfheart @NWBY
🥳 After hard work, the time has come: https://www.confetti-framework.com/docs/digging-deeper/commands

I also made a separate template for it. This allows you to use Confetti Commands if you just need to create a CLI tool:
https://github.com/confetti-framework/commands

**Reply by KinyaElGrande** [reply]
May I say I like the way that you have implemented it . Then everything just flows and the documentation to me is too informative I love it.

**Reply by reindert-vetter** [reply]
Thanks @KinyaElGrande!

Do you use something of Confetti at work? I'd love to hear what you're missing.

**Reply by KinyaElGrande** [reply]
Nop, currently I just use it with my personal side project. 
I don't have a job at the moment but in my previous Jobs, I used laravel a lot and **I love  Go** .

I haven't found proper time to look at Confetti been finishing up with my school this is the last month yeii . Soo from April I'll be free and I am going to interact with Confetti more and more , if something comes up I'll be raising it for sure.
